python_modules/libraries/dagster-azure/dagster_azure/adls2/resources.py

```diff
@@ -29,6 +65,38 @@
 }
 
 
+class ADLS2Resource(ADLS2BaseResource):
+    """Resource containing clients to access Azure Data Lake Storage Gen2.
+
+    Contains a client for both the Data Lake and Blob APIs, to work around the limitations
+    of each.
+    """
+
+    @property
+    @cached_method
+    def _raw_credential(self) -> Any:
+        if isinstance(self.credential, ADLS2Key):
+            return self.credential.key
+        elif isinstance(self.credential, ADLS2SASToken):
+            return self.credential.token
+        else:
+            return DefaultAzureCredential(**self.credential.kwargs)
+
+    @property
+    @cached_method
+    def adls2_client(self) -> DataLakeServiceClient:
+        return create_adls2_client(self.storage_account, self._raw_credential)
+
+    @property
+    @cached_method
+    def blob_client(self) -> BlobServiceClient:
+        return create_blob_client(self.storage_account, self._raw_credential)
+
+    @property
+    def lease_client_constructor(self) -> Any:
+        return DataLakeLeaseClient
+
+
 @resource(ADLS2_CLIENT_CONFIG)
```

Comment:
What precisely can we not map 1:1?

Reply:
Can you clarify what cannot be mapped 1:1?

Reply:
Ah sorry, left this in a comment below instead of in thread:

This is a limitation of the discriminated union setup, which requires that each element in the union is a Config type, e.g. not a raw string.

```python
class ADLS2SASToken(Config):
    credential_type: Literal["sas"] = "sas"
    token: str
```

w/ config dict
```python
{
  "storage_account": "foo",
  "credential": {
    "sas": {
      "token": "my_token"
    }
  }
}
```

Before, we would have the slightly more compact
```python
{
  "storage_account": "foo",
  "credential": {
    "sas": "my_token"
  }
}
```

Where the `sas` type is just a string, rather than a `Config` w/ string field.